source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.cc

```diff
@@ -0,0 +1,180 @@
+  // prevent the sample window from resetting until enough requests are gathered to complete the
+  // recalculation.
+  concurrency_limit_.store(1);
+  stats_.concurrency_limit_.set(concurrency_limit_.load());
+  recalculating_min_rtt_.store(true);
+
+  // Throw away any latency samples from before the recalculation window as it may not represent
+  // the minRTT.
+  absl::MutexLock ml(&latency_sample_mtx_);
+  hist_clear(latency_sample_hist_.get());
+}
+
+void GradientController::updateMinRTT() {
+  ASSERT(recalculating_min_rtt_.load());
+
+  // Reset the timer to ensure the next minRTT sampling window upon leaving scope.
+  auto defer =
+      Cleanup([this]() { min_rtt_calc_timer_->enableTimer(config_->min_rtt_calc_interval()); });
+
+  absl::MutexLock ml(&latency_sample_mtx_);
+  min_rtt_ = processLatencySamplesAndClear();
+  stats_.min_rtt_msecs_.set(
+      std::chrono::duration_cast<std::chrono::milliseconds>(min_rtt_).count());
+  recalculating_min_rtt_.store(false);
+}
+
+void GradientController::resetSampleWindow() {
+  // The sampling window must not be reset while sampling for the new minRTT value.
+  if (recalculating_min_rtt_.load()) {
+    return;
+  }
+
+  absl::MutexLock ml(&latency_sample_mtx_);
+  if (hist_sample_count(latency_sample_hist_.get()) == 0) {
+    return;
+  }
+
+  sample_rtt_ = processLatencySamplesAndClear();
+  concurrency_limit_.store(calculateNewLimit());
+  stats_.concurrency_limit_.set(concurrency_limit_.load());
+}
+
+std::chrono::microseconds GradientController::processLatencySamplesAndClear() {
+  const std::array<double, 1> quantile{config_->sample_aggregate_percentile()};
+  std::array<double, 1> calculated_quantile;
+  hist_approx_quantile(latency_sample_hist_.get(), quantile.data(), 1, calculated_quantile.data());
+  hist_clear(latency_sample_hist_.get());
+  return std::chrono::microseconds(static_cast<int>(calculated_quantile[0]));
+}
+
+uint32_t GradientController::calculateNewLimit() {
+  // Calculate the gradient value, ensuring it remains below the configured maximum.
+  ASSERT(sample_rtt_.count() > 0);
+  const double raw_gradient = static_cast<double>(min_rtt_.count()) / sample_rtt_.count();
+  const double gradient = std::min(config_->max_gradient(), raw_gradient);
+  stats_.gradient_.set(gradient);
+
+  const double limit = concurrency_limit_.load() * gradient;
+  const double burst_headroom = sqrt(limit);
+  stats_.burst_queue_size_.set(burst_headroom);
+
+  // The final concurrency value factors in the burst headroom and must be clamped to keep the value
+  // in the range [1, configured_max].
+  const auto clamp = [](int min, int max, int val) { return std::max(min, std::min(max, val)); };
+  const uint32_t new_limit = limit + burst_headroom;
+  return clamp(1, config_->max_concurrency_limit(), new_limit);
+}
+
+RequestForwardingAction GradientController::forwardingDecision() {
+  // Note that a race condition exists here which would allow more outstanding requests than the
```

Comment:
@mattklein123 the more I think about this, the more I disagree with dropping the compare/exchange loop. This eventually consistent approach would regularly violate the concurrency limit in systems that are shedding load. The eventual consistency would be totally fine if it was just the value of the concurrency limit, but we're dealing with stale outstanding request counts as well which I think is unacceptable. If I configure a concurrency of 1, or the system is in a minRTT recalculation window, I'd expect this to only let through a single request at a time.

We're throwing in unintuitive behavior that I think would confuse users of the filter to avoid using a well-documented function given by std::atomic<int>. If we're trying to avoid confusing developers with a lock-free approach using a CAS operation, then I'd suggest a more straight-forward mutex on `num_rq_outstanding_`.

Reply:
Now that I understand this code more, I think the CAS loop could spin for quite some time while waiting for requests to drain so I don't think that is acceptable? 

I definitely note what you are saying, but IMO this is an OK and simple way to start, and we can always iterate later. All of the other circuit breakers inside Envoy are eventually consistent so I don't see a problem with keeping this similar here also?

Reply:
I can agree that this is acceptable to start, so I won't die on this hill. However, I do want to clarify the first statement for my own understanding:

> Now that I understand this code more, I think the CAS loop could spin for quite some time while waiting for requests to drain so I don't think that is acceptable?

I just don't think I understand what you mean by request draining,. Do you mean waiting for the requests to be serviced so `num_outstanding_rq_` can drop below the limit? If so, the loop spins only if the `num_outstanding_rq_` is modified between the comparison against the limit and the increment via `compare_exchange_weak()`, which implies that there was forward progress on some other worker thread. When waiting for the outstanding requests to drop below the limit, we're just returning the 503 without causing another iteration of the loop.

Reply:
OK, I haven't read the code in enough detail to fully understand it. I will re-analyze once the rest of it is simpler per the rest of the review. Let's just drop this for now and we can come back to it later?